numpy/core/src/multiarray/dtypemeta.c

```diff
@@ -698,6 +735,39 @@ object_common_dtype(
     return cls;
 }
 
+static int
+fill_zero_object_strided_loop(
+        void *NPY_UNUSED(traverse_context), PyArray_Descr *NPY_UNUSED(descr),
+        char *data, npy_intp size, npy_intp stride,
+        NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyObject *zero = PyLong_FromLong(0);
+    PyObject **optr = (PyObject **)data;
+    while (size--) {
+        Py_INCREF(zero);
+        *optr++ = zero;
```

Comment:
In order to correctly support unaligned memory (for structured dtypes) at least.  We probably need to use a `memcpy` here unfortunately.  (and a `char *` pointer)

Reply:
Ah my first pass over `PyArray_FillObjectArray` made me think that this is what it was doing, but it's actually doing a `memcpy`. Thanks for this catch also.